Add Process Mining Exporter and exec/github.R

Ssunoo2 opened the pull request:
Script to call github refresh from the command line

Ssunoo2 committed:
Create github.R

carlosparadis committed:
fix github actions

carlosparadis commented on .github/workflows/R-CMD-check.yml:
Make sure you revert this or sync to master

carlosparadis commented on .github/workflows/test-coverage.yml:
same on revert to master

carlosparadis commented on exec/github.R:
```diff
@@ -0,0 +1,100 @@
+  save_path_issue_or_pr_comments <- paste0(save_path,"/issue_or_pr_comment/")
+  # Path you wish to save all raw data. A folder with the repo name and sub-folders will be created.
+  owner <- conf[["issue_tracker"]][["github"]][["owner"]] # Has to match github organization (e.g. github.com/sailuh)
+  repo <- conf[["issue_tracker"]][["github"]][["repo"]] # Has to match github repository (e.g. github.com/sailuh/perceive)
```
hardcoded get

carlosparadis reviewed (changes requested):
pointing out here the get and other files ahead of time, same issue as @daomcgill

RavenMarQ committed:
Update github.R

To standardize github.R to match mailinglist.R executable. Also updated implementation to utilize the configuration get functions.

connorn-dev committed:
i #341 moved from PR 295 - Updated Vignette

Includes steps to:
- download and parse data
- generate start and end activities and process tree

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

connorn-dev commented:
@carlosparadis just added the code from Mbox. I kept github.R separate from my changes and added the vignette.

carlosparadis commented:
@connorn-dev sounds good i will go ahead and add a few quick notes as review to your code here

carlosparadis commented on exec/ghevents.R:
```diff
@@ -0,0 +1,89 @@
+if (arguments[["download"]] & arguments[["help"]]) {
+  cli::cli_alert_info("Downloads Github Events using R/github.R")
+} else if (arguments[["download"]]) {
+  owner <- arguments[["<owner>"]]
```
Should use get() functions instead of direct access. See the cheat sheet associated to it for rationale on why.

connorn-dev replied:
So for example, to get the owner of the desired Github I should use "get_github_owner()" and rely on the configuration file for the users input?

carlosparadis replied:
The path to the configuration file should be passed to your /exec. However, to load the config from the path, and the subsequent accesses to get field information should use get() functions. The functions exist on R/config.R. if it doesn't exist, you should define new ones. See "accessing configs from Notebooks". That also applies to configs: https://github.com/sailuh/kaiaulu_cheatsheet/blob/main/cheatsheets/project-analysis-cheatsheet.pdf 

https://github.com/sailuh/kaiaulu/blob/c5922608d8a88287767a523bc9194ee9ae032b0b/exec/mailinglist.R#L89-L90

Not all the code on PR is fully reviewed (I see Dao missed the config get to load the file), but it exists: 

https://github.com/sailuh/kaiaulu/blob/ac522b6fd675230da905c0b54b241904e6c76705/R/config.R#L21-L30

See the cheatsheet linked above why we enforce the get() this was a major milestone for the Fall'24 group : )

carlosparadis commented on vignettes/event_process_mining.Rmd:
````diff
@@ -0,0 +1,70 @@
+### Install Reticulate
+
+```{r}
+install.packages("reticulate")
````
You should not use reticulate. This is a strong dependency on Kaiaulu side. Rather, it is the Python script that should call this exec script.

carlosparadis reviewed (changes requested):
Here's some revision notes. Let me know if you have questions. I suggest you reply to the comment in question you want to iterate rather than an entirely separate comment so things are in context to the line of code.

connorn-dev committed:
i #341 Updated Introduction of Vignette

- Provides more info on process mining

connorn-dev committed:
i #341 Created Config File

connorn-dev committed:
i #341 Updated Config file & Vignette

connorn-dev committed:
i #341 Changed Conf File

connorn-dev committed:
i #341 Switched to get functions

- Switched CLI to config file
- Switched to get functions from config file
- Added R/config.R to branch
- Added get csv path to R/config.R

connorn-dev committed:
i #341 Added to github_api_showcase

- Added usage of CLI to github_api_showcase
- Changed to using Kaiaulu config for python notebook

connorn-dev committed:
i #341 Cleaned up vignette deleted config

- Cleaned up vignettes content
- Deleted process_mining.yml config file
- Deleted get_pm_csv_path function from R/config

connorn-dev committed:
i #341 Added NEWS.md and Contributor

- Updated NEWS.md with one sentence update for ghevents.R
- Contributor added to DESCRIPTION

carlosparadis commented on vignettes/event_process_mining.Rmd:
Hmm. Up to this point notebooks document the API, not the execs. I will need to give more thought on this in the future.

carlosparadis reviewed (changes requested):
Looks fine on this side. Commits are blended to older PR so I can only do a cursory overview. Please disregard the request change, this is a note to self when I work on the merge over the Summer.

connorn-dev committed:
i #341 cleaned up wording of vignette

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

carlosparadis commented on exec/ghevents.R:
```diff
@@ -0,0 +1,101 @@
+  github_events.R parse help
+  github_events.R parse <config_path> <output_file>
+  github_events.R (-h | --help)
+  github_events.R --version
```
Could you check the other exec to see which should we stick with for consistency? E.g. if the others are all github_* then it may be best we move ghevents to github_events.R throughout the code.

carlosparadis commented on exec/github.R:
```diff
@@ -0,0 +1,144 @@
+    save_path_issue_or_pr_comments <- path.expand(get_github_issue_or_pr_comment_path(conf, project_key))
+
+    dir.create(save_path_issue_or_pr_comments)
+    github_api_iterate_pages(token,gh_response,
```
Isn't it defined on line 64? This is the entirety of the code block after it is defined:

https://github.com/sailuh/kaiaulu/blob/8b0582539bbb18b8d4d2f2e410abe554b4c5acba/exec/github.R#L64-L90

geraldmjhuff replied:
gh_response in line 64 was defined within the scope of the 1st if block. Since gh_response was called in line 87 in the 3rd if block, it doesn't see gh_response. 

I tested both scenarios where I'd add gh_response <- github_api_project_issue_or_pr_comments(owner, repo, token) in line 85 and one without it. The one without it gave me this error: Error: object 'gh_response' not found
Execution halted

carlosparadis commented on exec/ghevents.R:
```diff
@@ -0,0 +1,101 @@
+  cli::cli_alert_info("Parsing Github issue events from {input_dir}")
+
+  all_issue_event <- lapply(list.files(input_dir, full.names = TRUE), read_json)
+  all_issue_event <- lapply(all_issue_event, github_parse_project_issue_events)
```
Looking at the requirement list, I see: 

https://github.com/sailuh/kaiaulu/blob/8b0582539bbb18b8d4d2f2e410abe554b4c5acba/exec/github.R#L12C17-L12C24 

As Kaiaulu is an R package, all files in Kaiaulu and its API should be loaded using this command. Did the code not rn for you?

carlosparadis commented on exec/github.R:
```diff
@@ -0,0 +1,144 @@
+require(knitr,quietly=TRUE)
+require(magrittr,quietly=TRUE)
+require(gt,quietly=TRUE)
+
```
Looking at the requirement list, I see: 

https://github.com/sailuh/kaiaulu/blob/8b0582539bbb18b8d4d2f2e410abe554b4c5acba/exec/github.R#L12C17-L12C24 

As Kaiaulu is an R package, all files in Kaiaulu and its API should be loaded using this command. Did the code not rn for you?

geraldmjhuff replied:
I see my mistake; I haven't properly installed the packages and therefore I had to source both scripts. After installing the packages, it works as intended and sourcing those two scripts are no longer necessary.

carlosparadis replied:
No worries, it is a common mistake!

geraldmjhuff commented on vignettes/github_api_showcase.Rmd:
```diff
@@ -55,22 +53,20 @@ token <- scan("~/.ssh/github_token",what="character",quiet=TRUE)
 
```
In save_path <- path.expand(conf[["issue_tracker"]][["github"]][["replies"]]) , tries to access "replies" in kaiaulu.yml but that key does not exist.

carlosparadis replied:
If the get() functions already exist, I agree you can proceed with this!

geraldmjhuff reviewed (commented):
@carlosparadis Made comments on the event_process_mining.Rmd and github_api_showcase.Rmd notebooks

carlosparadis commented on vignettes/event_process_mining.Rmd:
```diff
@@ -0,0 +1,59 @@
+command <- 'Rscript exec/ghevents.R download conf/process_mining.yml --token_path=~/.ssh/github_token'
+system(command)
+
+command <- 'Rscript exec/ghevents.R parse ../output ../output.csv'
```
Thank you for catching this. Unfortunately, this notebook implementation is incorrect. So I am glad you found it and asked me about it to avoid also doing the same. You see, the R notebook in Kaiaulu purpose is to explain to the user how to use Kaiaulu API (functions in R/) to write their own analysis. So, although the code here is functional, it is not complying to Kaiaulu architecture. There is also no reason to use the script when the user can just use the R function, it is an R notebook afterall.

The place which the exec should be used is *outside* Kaiaulu. This generally makes more sense if you are using another language, in which case you will not have the ability to call an R function directly or it is less desirable. For example, @austindang67 would want to issue a command line call to Kaiaulu to get the data he needs. This also keeps his side of the code clean on the frontend (the only responsability of the frontend is to request data from Kaiaulu, not preprocess it, etc). 

Does that make sense? So the fix here would be to go to the script exec/ghevents.R and bring in the code block that is calling functions there into the notebook. In the notebook you may have a few code blocks and english around the code blocks explaining it. 

Or is this code just cited here to acknowledge it exists and the rest of the notebook goes API function-wise?

geraldmjhuff replied:
@carlosparadis I've rewritten the notebook to no longer mention any CLI, explained the use case of the project config file and why it's necessary, and shifted to an API-only related text notebook that explains how to download and parse repo's using functions from ghevents.R. Process_mining is mentioned to let the user know a use case for the parsed saved data.

This is the current code that is working as intended in my codespace: 

---
title: "Event Process Mining"
output: 
  html_document:
    toc: true
    number_sections: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Event Process Mining}
  %\VignetteEncoding{UTF-8}
---

# Introduction

This notebook demonstrates how to use the **Kaiaulu R API** functions defined in `ghevents.R`  
to download and parse GitHub issue event data using the GitHub API.  
This data can be used for various analysis and use cases. A possible use case is process mining analysis, to see an example for process mining refer to this [repo](https://github.com/sailuh/process_mining).

## Requirements

- **GitHub Access Token**  
  Used for authentication when accessing the GitHub REST API.

- **Kaiaulu**  
  Provides the API functions implemented in `ghevents.R` for retrieving and parsing GitHub event data.

## Project Configuration File

To simplify setup and ensure consistency, Kaiaulu uses a configuration file (e.g., `kaiaulu.yml`)  
that defines project-specific parameters such as the GitHub owner, repository name, and data storage paths.  

This configuration file enables the API functions to locate repositories and determine where to store downloaded and parsed data.  
The content that interests us for this notebook is primarily the `issue_event` download path.  

By parsing this configuration file at the start, the notebook clearly communicates **all information needed to run the API functions**:  

- Which GitHub organization (`owner`) and repository (`repo`) to query  
- Where to save raw JSON issue event data (`issue_event`)  
- Which token to use for authentication (`token_path`)  
- Where to export the processed CSV (`output_file`)  

This ensures that the notebook workflow is fully reproducible and that users do not need to manually supply repository details or paths.

An example configuration snippet from `kaiaulu.yml`:

    ```        
    github:
        project_key_1:
          # Obtained from the project's GitHub URL
          owner: sailuh
          repo: kaiaulu
          # Download using `download_github_comments.Rmd`
          issue_or_pr_comment: ../../rawdata/kaiaulu/github/sailuh_kaiaulu/issue_or_pr_comment/
          issue: ../../rawdata/kaiaulu/github/sailuh_kaiaulu/issue/
          issue_search: ../../rawdata/kaiaulu/github/sailuh_kaiaulu/issue_search/
          issue_event: ../rawdata/kaiaulu/github/sailuh_kaiaulu/issue_event/
          pull_request: ../../rawdata/kaiaulu/github/sailuh_kaiaulu/pull_request/
          commit: ../../rawdata/kaiaulu/github/sailuh_kaiaulu/commit/

    ```
## Download and Parse GitHub Issue Event Data

The following code demonstrates the full workflow of the Kaiaulu R API to:

1. Load the project configuration file to obtain necessary parameters (owner, repo, paths, and token).  
2. Download all GitHub issue event data for the specified repository.  
3. Parse the raw JSON files into a clean tabular format.  
4. Export the processed data to a CSV file for downstream analysis, such as process mining.

```{r}
library(kaiaulu)
library(data.table)
library(jsonlite)

conf <- parse_config("conf/kaiaulu.yml")
owner <- get_github_owner(conf, "project_key_1")
repo <- get_github_repo(conf, "project_key_1")
save_path <- get_github_issue_event_path(conf, "project_key_1")
token_path <- "~/.ssh/github_token"
output_file <- "../output.csv"

token <- scan(token_path, what="character", quiet=TRUE)
if(!dir.exists(save_path)) dir.create(save_path, recursive=TRUE)

# Download GitHub issue events 
gh_response <- github_api_project_issue_events(owner, repo, token)
github_api_iterate_pages(token, gh_response, save_path, prefix="issue_event")

#  Parse downloaded JSON files 
json_files <- list.files(save_path, full.names=TRUE)
all_issue_event <- lapply(json_files, read_json)
all_issue_event <- lapply(all_issue_event, github_parse_project_issue_events)
all_issue_event <- rbindlist(all_issue_event, fill=TRUE)
all_issue_event[, issue_body := NULL]

write.csv(all_issue_event, output_file, row.names=FALSE)
```

### Next Steps: Process Mining Analysis

The processed GitHub issue event data is now saved in `output.csv`.  
If you want to perform process mining analysis, you can use the [Process Mining Notebook](https://github.com/sailuh/process_mining),  
which relies on Python and the `pm4py` package to analyze issue event data.

carlosparadis replied:
@geraldmjhuff, I am actually now confused about what is going on here. Can you do this for me: Ping me on the issue associated to this task (so I can get context about it), and then paste this comment (https://github.com/sailuh/kaiaulu/pull/301#discussion_r2409672770) so  I can follow-up. With that said, I did not intend to suggest a new issue or PR. I assumed whatever notebook you posted here was already part of this PR/branch and you pasted it on a comment on the PR instead of just making a commit changing it. Regardless, please post on the issue this comment for us to iterate there since this is relevant for memory keeping.

For future reference: Discussion on PR will rarely be seen again (since it is a very specific need to go back to a code review via git blame), but the issues will be important in the future.

Thanks!

geraldmjhuff committed:
Updated exec/github.R documentation

Refresh commands now above parse commands in the case user wants to run these in a linear order.

geraldmjhuff committed:
Added missing gh_response in exec/github.R 

Added gh_response <- github_api_project_issue_or_pr_comments(owner, repo, token) which was originally missing.

geraldmjhuff committed:
Replaced config access lines w/ get_* functions in github_api_showcase.Rmd

Replaced multiple hardcoded config access lines with get_* functions that functionally yield same results.

geraldmjhuff committed:
Removed CLI related text in github_api_showcase.Rmd

Removed CLI related documentation and code chunks to follow Kaiaulu notebooks API only architecture.

geraldmjhuff committed:
Replaced CLI with API calls in event_process_mining.Rmd

CLI-related text and documentations were replaced with API calls to follow Kaiaulu's API only notebook architecture.

geraldmjhuff committed:
Fixed merge conflicts

geraldmjhuff committed:
Updated vignettes/event_process_mining.Rmd

Added workspace reset, set seed, and required libraries R commands. Corrected YAML access.

geraldmjhuff committed:
Updated exec/github.R

Added additional parameter <project_key> when using parse command to accommodate YAML access. Parse code blocks uses project_key dynamic access in get_github_*_path() but did not have a project_key assignment, and therefore could not run, this change fixes that.

carlosparadis commented:
Thank you for conducting the investigation! If this is the case then, are we in agreement that:

```r
parsed_response[["body"]] <- api_response[["body"]]
```

will be wrapped in an if(!is.null) statement, with a comment above it stating this controls for the `No description provided` cases? I would not add the if to the rest, as we would rather diagnose at the time the cause of the issue (especially since you can't post a comment without any description, as you can with an issue).  


```r
github_parse_project_issue <- function(api_responses){
  parse_response <- function(api_response){
    parsed_response <- list()
    parsed_response[["issue_id"]] <- api_response[["id"]]
    parsed_response[["issue_number"]] <- api_response[["number"]]
    parsed_response[["html_url"]] <- api_response[["html_url"]]
    parsed_response[["url"]] <- api_response[["url"]]
    parsed_response[["created_at"]] <- api_response[["created_at"]]
    parsed_response[["updated_at"]] <- api_response[["updated_at"]]
    parsed_response[["state"]] <- api_response[["state"]]
    parsed_response[["issue_user_login"]] <- api_response[["user"]][["login"]]
    parsed_response[["author_association"]] <- api_response[["author_association"]]
    parsed_response[["title"]] <- api_response[["title"]]
    parsed_response[["body"]] <- api_response[["body"]]

    parsed_response[["labels"]] <- api_response[["labels"]]
    if(length(parsed_response[["labels"]]) > 0){
      parsed_response[["labels"]] <- stringi::stri_c(sapply(parsed_response[["labels"]],"[[","name"),collapse = ",")
    }else{
      parsed_response[["labels"]] <- NA_character_
    }

    parsed_response <- as.data.table(parsed_response)

    return(parsed_response)
  }
  rbindlist(lapply(api_responses,parse_response),fill=TRUE)
}
```

geraldmjhuff committed:
Updated conf/kaiaulu.yml

Fixed issue_event and commit folder access. Previous commit 76b333f1528faa83f30911dc7fd067993f629aab accidentally changed these lines. This commit was made instead of reverting aforementioned commit due to having multiple file changes within one commit.

geraldmjhuff commented:
@carlosparadis Hmm that is right, the code does just work with:
```R
# For issues with "No description provided" in the top comment box; if body is NULL or empty, assign NA
parsed_response[["body"]] <- 
  if (length(api_response[["body"]]) == 0) {
    NA_character_
  } else {
    api_response[["body"]]
  }
```
My mistake, I've autopiloted doing the `is.null(api_response[["body"]]` just because since the JSON has a null value, i'd just add the NA in place of it and think it would work, which is clearly not the case, as length(api_response[["body"]]) == 0 itself accommodates both null and zero-length vectors. Similar to how it is done in:
```R
parsed_response[["labels"]] <- api_response[["labels"]]
    if(length(parsed_response[["labels"]]) > 0){
      parsed_response[["labels"]] <- stringi::stri_c(sapply(parsed_response[["labels"]],"[[","name"),collapse = ",")
    }else{
      parsed_response[["labels"]] <- NA_character_
    }
```

carlosparadis commented:
Please proceed with the commit so it is consistent with the rest of the codebase. Could you just check with a print() to check what the json library we use is parsing as? For future reference. 

Also, can you confirm if you are interested in trying the unit test? Otherwise I will make another issue and defer for the future. Either way is fine for me, just checking on your interests given your timeline to finish the filter and notebook.

As far as the merge is concerned, go ahead and do it, although I do think we will need to do one more once I get a chance to merge the PR for Pull Request API on main since the notebooks were refactored in name and slightly in content. We can worry about later on (or I will worry about it on the occasion).

Thanks for this!

geraldmjhuff committed:
Fixed handling of 'No description provided' in issue body

github_parse_project_issue() and github_parse_project_pull_request() now handles issues with 'No description provided' in their top comment box.

geraldmjhuff committed:
Merge branch 'master' into 298-github-refresh-script

geraldmjhuff committed:
Added unit test for GitHub issue with NULL description

- Added example_github_issue_no_description() to create a synthetic GitHub issue with NULL body.
- Added unit test to confirm github_parse_project_issue() converts NULL to NA

geraldmjhuff commented:
@carlosparadis 
Yes this PR should be ready for review, and I don't believe anything from sentiment should be here instead.

And yes I've also applied the minor adjustments above before making the commit.

carlosparadis commented:
@geraldmjhuff As you can see, I finally merged #323 to master. Could you do the git merge locally on your computer from master to see if your fixes conflict with what was pushed there? This would not only save me an immense amount of time, but it would also ensure your findings and fixes in this PR are not lost if I try to merge myself. 

Note vignettes/github_api_showcase.Rmd still exists, the file was just renamed to be download_events.Rmd. 

R/github.R and the associated parse_replies is likely where it affects you.

geraldmjhuff committed:
Merge masters to this branch

geraldmjhuff committed:
Update NEWS.md to remove Conflict strings

geraldmjhuff committed:
Updated R/github.R to remove Conflict strings

geraldmjhuff committed:
Updated R/github.R, exec/github.R

Updated R/github.R to match masters, then implemented NA checks in
multiple parse functions.

geraldmjhuff committed:
Merge branch '298-github-refresh-script' of https://github.com/sailuh/kaiaulu into 298-github-refresh-script

geraldmjhuff commented:
@carlosparadis 

I've made the merge from masters, exec/github.R and exec/ghevents.R works as intended, this PR should be ready for review.

carlosparadis committed:
i #298 #341 Add Process Mining Exporter

Add process mining exporter and some fixes to exec/github.R

Signed-off-by: Carlos Paradis <carlosviansi@gmail.com>

carlosparadis commented:
Had entirely forgotten this PR had evolved to have the Process Mining exporter tied to sailuh/process_mining. Cleaned up some of the associated Notebook and edited the PR.

carlosparadis committed:
Add missing rd

Signed-off-by: Carlos Paradis <carlosviansi@gmail.com>

carlosparadis committed:
Fix NEWS.md

ghevents.R was not merged to github.R fixes news.

Signed-off-by: Carlos Paradis <carlosviansi@gmail.com>

carlosparadis committed:
Add Connor as contributor

Signed-off-by: Carlos Paradis <carlosviansi@gmail.com>
